### ISSUE TITLE
Migrate to Gtk3

### ISSUE THREAD

**Issue by jplatte** [squentin/gmusicbrowser#57]
The seemingly unknown perl-gtk3 module is available at https://git.gnome.org/browse/perl-Gtk3/

**Comment by genodeftest** [comment]
[According to Michael Catanzaro](https://blogs.gnome.org/mcatanzaro/2016/02/01/on-webkit-security-updates/), a developer of WebKitGtk, WebKit 1 Gtk bindings (in this case Gtk2 bindings) miss many security updates and are discouraged to use. Please stop using WebKit1 Gtk bindings. This requires porting to Gtk3, since WebKit2 is only available for Gtk3.

**Comment by remicollet** [comment]
FYI, see [RHBZ #1375826] perl-Gtk2-Webkit will be removed soon from Fedora repository (in F26, in a few months), making impossible to support gmusicbrowser there.

**Comment by genodeftest** [comment]
> FYI, see RHBZ #1375826 perl-Gtk2-Webkit will be removed soon from Fedora repository (in F26, in a few months), making impossible to support gmusicbrowser there.

Actually, it will be removed in Fedora 27 which will be released in about one year.

**Comment by squentin** [comment]
> FYI, see RHBZ #1375826 perl-Gtk2-Webkit will be removed soon from Fedora repository (in F26, in a few months), making impossible to support gmusicbrowser there.
webkit is only used by one plugin, it has never been an important part of gmb, so its removal shouldn't be a big issue.

**Comment by intrigeri** [comment]
FYI libgtk2-perl [won't be part of Debian Bullseye](https://bugs.debian.org/912860) (likely release: mid-2021). On the gmb Debian front, this is tracked as https://bugs.debian.org/912882.

**Comment by jspricke** [comment]
I started a port to GTK3 here: https://github.com/jspricke/gmusicbrowser/tree/gtk3
The main windows is there and you can control the player, but the songs are not listed. The last commit contains all stuff that is commented out and the command I use for testing. Maybe someone can help with this? I'm also in #gmusicbrowser, if you want to discuss. (Note that I will rebase the commits from time to time.)

**Comment by genodeftest** [comment]
Personally, I prefer [Exaile](https://www.exaile.org/) with its github repo [exaile/exaile](https://github.com/exaile/exaile). That player recently got its initial Gtk3 release published. You may want to take inspiration from its porting procedure. Anyway, I fear the GObject Introspection support for Perl will probably be worse than its python bindings so expect some issues there. And I guess this will be quite some work to do, especially on a barely maintained project such as gmusicbrowser.

Disclaimer: I sometimes contribute to Exaile ;)

**Comment by carl-di-ortus** [comment]
Just updated to Ubuntu 19.10 and to my surprise - gmusicbrowser is gone! This needs to be addressed really urgent.

**Comment by jspricke** [comment]
As discussed in #gmusicbrowser, there are now two branches:
https://github.com/jspricke/gmusicbrowser/tree/gtk3 (containing the needed patches)
https://github.com/jspricke/gmusicbrowser/tree/gtk3-testing (containing hacks for testing, will be rebased onto the gtk3 if there are changes)

**Comment by bor** [comment]
Just point out the answer from Quentin about gtk3 status
https://github.com/squentin/gmusicbrowser/issues/149

**Comment by yennor** [comment]
I've got nothing productive to say. But welcome back squentin . I'm happy to hear the best player ever has got a future :-)

**Comment by squentin** [comment]
Just a quick update, what is left to be ported to gtk3:

- list-related stuff (in gmusicbrowser_list.pm haven't touched the file yet)
- plugins (should be easy)
- drag and drop
- icons still need more work
- box sizing in layouts, (particularly affect the album picture)
- a few problems that will need more work but that can wait
- various small things that are not really important

**Comment by squentin** [comment]
Sorry for the lack of updates.

The port is mostly finished, including plugins, I even resurrected the old webcontext plugin after noticing that there was introspection data for webkit which is now all that is needed to use a library in perl.

There is still some issues with sizing and icons, and I need to do some cleanups. I'll probably push it to git in a week or two (no promises).

Another issue is that the SongList is a bit slower than he used to be, in particular during its creation and filling/emptying. Not sure I can improve it, but it can wait. At least the SongTree is still fast and now uses real TreeView column headers that can be dragged and dropped, so it's a good alternative.

**Comment by squentin** [comment]
I've just pushed it to master. (the date of the commit says February because I forgot to update the commit date after I squashed all the messy commits into one)

- as I mentioned before the SongList is very slow when loading/unloading large number of songs, you can replace it your layout with a SongTree (remove all the groups in "edit grouping" to get a simple list)

- the save file (gmbrc) is compatible between the old and new version, so you can try the new version and go back if you encounter a problem.

- there are some sizing issues due to not being able to port correctly my box size allocator for layouts, but I think it should work fine with some layout options, I've only updated layout options of a few I use for now. So you might need to change some options in your layouts. In particular 'expand_to_ratio' and 'expand_weight' options have been removed and maxwidth only work for labels.

- icons mostly work but are still a mess, and need to be reworked

**Comment by justperfection** [comment]
Thanks for the update and hard work.
Please,
- Move to the svg for icons since it is much better for HiDPI.
- Allow icon themes to have smallclosetab. Currently all themes using the main smallclosetab.png.
- Give the option to switch to the dark variant since we can do that in gtk3.
- Implement CSD if it is possible.

**Comment by carl-di-ortus** [comment]
@squentin I would like to note that the next Debian freeze is in just over a half year. It would be an achievement to [get gmusicbrowser back to it](https://tracker.debian.org/pkg/gmusicbrowser), allowing stable users to just continue using with the next release. More popular side-quests like Ubuntu would eventually catch up as well. I reckon that player got removed only because of old dependencies, that will not make into next release. So if GTK is now updated, maybe some outdated perl packages as well, then it's time for a version bump for gmusicbrowser, and submit a new package. Do you need some help with maintaining the Deb package? Maybe maintaining this repo (there are pull requests back from 2013)?

@justperfection all your points could be a separate feature requests. I would like to see a more quick updated release right now and handle additional features later on.

**Comment by newhoa** [comment]
@squentin You mention an icon rework. I don't know if you meant you wanted to recode the way they work or what. But I did [fork this current build](https://github.com/newhoa/gmusicbrowser) and modified almost all the icons to use system theme icons instead of the included icons. Icons are a little more consistent, easier to see due to using symbolic icons, are SVG in modern icons themes (for better HiDPI scaling), some missing icons were added, and it should now match any user selected icon theme.

Not sure if this is a change you want to include, but thought I'd give it a try. [Pull request is here](https://github.com/squentin/gmusicbrowser/pull/177).

Thanks for everything! This is by far my favorite music player! So glad this is working with Gtk3 now!

**Comment by yennor** [comment]
I've just checked out the current master. It seems to be working on debian. :-).
@squentin
Just a question, do you already want bug reports on the new version (like it's not possible to add a new song to a saved list). Or are you still digging around a lot in the code and prefer to officially do a new release before we start filling the bug tracker ;-) ?

**Comment by just-a-bug-reporter** [comment]
How far we are from GTK3 port to get completed?

**Comment by squentin** [comment]
Sorry for the lack of replies.

> How far we are from GTK3 port to get completed?
It can be considered almost complete already, the only things left are the icons issue, updating some layouts, and fixing some minor-but-annoying bugs, I use it daily.


>  Just a question, do you already want bug reports on the new version (like it's not possible to add a new song to a saved list). Or are you still digging around a lot in the code and prefer to officially do a new release before we start filling the bug tracker ;-) ?
I didn't reply but I fixed that bug in git after you reported it.
As you might have noticed, I haven't worked much on gtk3 problems lately. There is a few known bugs, you are welcome to report other bugs.

Known bugs (in git, a few bugs have been fixed since last release):
- icons: with no gmb icon theme, play/next/previous icons are not shown
- icons: gmb icon theme only working for some icons and requires restart
- icons: display the missing icon (instead of nothing) in some places for fields like labels when no icon is defined for a value
- icons: a few dialog buttons have an icon but no text
- some panels do not display properly when they don't have their full width
- SongList columns sometimes lose their saved width somehow
- desktop widgets set as insensitive are sometimes not insensitive
- some layouts have size problems and need to be updated

**Comment by justperfection** [comment]
For icons, I think you should remove icon themes and use symbolic icons instead.
It will simplify the process, works better in Hi-dpi and it is recommended by [GNOME Human Interface Guidelines](https://developer.gnome.org/hig/stable/).

btw, settings need [18px padding](https://developer.gnome.org/hig/stable/visual-layout.html.en).

**Comment by squentin** [comment]
> For icons, I think you should remove icon themes and use symbolic icons instead.
> It will simplify the process, works better in Hi-dpi and it is recommended by [GNOME Human Interface Guidelines](https://developer.gnome.org/hig/stable/).
I will make symbolic icons an option, maybe even the default. But I still want to make it easy for users to replace some icons, I think it's important for icons that users are seeing all day long. I don't remember all the details, but It wouldn't help much anyway, I would still have to rename all the icons, and still have to use a separate system for user-definable icons like for labels.

> btw, settings need [18px padding](https://developer.gnome.org/hig/stable/visual-layout.html.en).
I'll try it, I guess there is a lot of small things like that. At least I fixed the alignment of checkboxes in the Library tab last month :)
You could open a detailed bug about the most important ones if you want.
By the way, I was recently wondering about the button order of dialogs (ok, cancel), I may make it an option as there seem to be 2 standards, and it's not much work.

**Comment by justperfection** [comment]
@squentin 
I know you did most of the port and you are alone on this project and fixing these things can be very time consuming. Since the final GTK3 port isn't ready yet, I recommend these solutions to speed up the process :

- Getting donation to help you work more hours on this project.
- Getting help from community to work on the port.

I think announcing it on Reddit can help you with both.

btw, I'm not a Perl developer but I know GTK4. If you need any help count me in. I can also help you to spread the words on my YouTube Channel. Just give me the signal.

You can also ask questions directly from [very friendly GTK developers](https://gnome.riot.im/#/room/#gtk:gnome.org).

**Comment by justperfection** [comment]
Can you please fix these issues so the gtk3 version is more usable?

- When filter label under the song cover cannot fit the size we will have size issue:

  ![gmusicbrowser gtk3](https://user-images.githubusercontent.com/66198705/147817146-2aa30d50-be6c-4bae-ab49-65b52ea64473.png)

- Some icons not showing up
  
  ![gmusicbrowser gtk3](https://user-images.githubusercontent.com/66198705/147817299-7a9a6b99-7777-4e56-ad2b-9a4598731614.png)

- Some album covers not showing up while it is working on gtk2 version.

- Menu button in shimmer desktop cannot be open again after closing it. I mean this button:
  
  ![gmusicbrowser gtk3](https://user-images.githubusercontent.com/66198705/147817876-a7a9d8f0-4330-43df-b527-0db7da276c19.png)

**Comment by ig3** [comment]
I cloned master to my system running Debian bullseye with xfce4 and it is running nicely without me having to install any additional packages. I could leave it at this, but a Debian package that I and others could install would be nice.

@squentin Would you consider a new release (maybe 1.1.99.2) including changes since 1.1.99.1?

Something that could be a basis for a new Debian package. From my perspective, the current state is better than no package, but maybe there are issues I'm not aware of. I haven't reviewed all the open issues.
